### PR TITLE
[Port] StepTriggerGroup From WhiteDream

### PR DESCRIPTION
**Pull request by VMSolidus**
# Description

This is a port of https://github.com/WWhiteDreamProject/wwdpublic/pull/53 from White Dream. This PR improves the StepTriggerImmune component by making it operate on a more granular Blacklist system, such that StepTriggerImmune entities can further clarify via prototypes which kinds of floor traps they are immune to, such as landmines/mousetraps, and not have blanket immunity to everything. Because it turns out things like Lava and Soap also were caught by the immunity, when really we just wanted Harpies & Felinids to not trigger landmines. 

<details><summary><h1>Media</h1></summary>
<p>

> # Описание
> Необходимо настроить модификатор урона, чтобы IPC не получали урон от осколков стекла.
> 
> Иммунитет StepTriggerImmuneComponent доработан. Теперь имеются несколько типов (types): Lava - тип тайла, наступив на который появляется урон. Это собственно лава и LiquidPlasma Landmine - мины. Chasm - дырка в карте, куда можно провалиться Mousetrap - Мышеловка SlipTile - Все, что должно подскальзывать игроков, имеющее размер тайла SlipEntity - Все, что должно подскальзывать игроков, имеющее развер энтити. Разделено для баланса. Самые ловки могут игнорировать мелкие предметы (энтити), т.е. уворачиваться от них. Но большие по площади вещи (тайлы по типу разлитой воды, бананиума) просчитываются отдельно.
> 
> # Изменения
> * [x]  Улучшить StepTriggerSystem (Immune)
> * [x]  Добавлены типы триггера. - Lava Landmine Shard Chasm Mousetrap SlipTile SlipEntity
> * [x]  Исправить осколки у IPC
> * [x]  Исправить отсутствие урона от лавы и падение в дыры у фелинидов и гарпий.
> 
> 🆑 Hell_Cat
> 
> * Feature: StepTriggerSystem is improved | Улучшена StepTriggerSystem
> * fix: IPC: Immunity for shards and SpiderWeb | Иммунитет осколкам.
> * fix: Felinid | Фелиниды : Immunity for Shard Landmine Mousetrap SlipEntities | Иммунитет для осколков, жидкости, мин, мышеловок, мыла и бананов.
> * fix: Harpy | Гарпия : Immunity for Shards Landmine Mousetrap | Иммунитет для осколков, жидкости, мин и мышеловок.
> * fix: Mice | Мыши : Don't blow up on landmines | Мыши не подрываются на минах.

</p>
</details>

# Changelog

:cl: Hell_Cat
Feature: StepTriggerSystem has been improved with new StepTriggerGroups. Additionally, the StepTriggerImmune component now allows declaring for specific StepTriggerGroups for a given entity to be immune to. Some examples may be, Felinids, Mice, and Harpies being unable to set off Landmines.

### CODE REVIEW

File: Resources/Prototypes/Entities/Effects/chemistry_effects.yml [hunk 1]
```diff
@@ -77,6 +77,9 @@
     animationState: foam-dissolve
   - type: Slippery
   - type: StepTrigger
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Effects/chemistry_effects.yml:80]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Entities/Mobs/NPCs/animals.yml [hunk 2]
```diff
@@ -1689,6 +1689,10 @@
   - type: FireVisuals
     sprite: Mobs/Effects/onfire.rsi
     normalState: Mouse_burning
+  - type: StepTriggerImmune # WD EDIT START
+    whitelist:
+      types:
+      - Landmine # END
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Mobs/NPCs/animals.yml:1695]:
```suggestion
  - type: StepTriggerImmune
    whitelist:
      types:
      - Landmine
```

File: Resources/Prototypes/Entities/Mobs/Player/ipc.yml [hunk 3]
```diff
@@ -111,6 +111,10 @@
   - type: OfferItem
   - type: LayingDown
   - type: Carriable
+  - type: StepTriggerImmune #WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Mobs/Player/ipc.yml:114]:
```suggestion
  - type: StepTriggerImmune
```

File: Resources/Prototypes/Entities/Mobs/Species/harpy.yml [hunk 4]
```diff
@@ -123,6 +123,11 @@
     - GalacticCommon
     - SolCommon
   - type: StepTriggerImmune
+    whitelist: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Mobs/Species/harpy.yml:126]:
```suggestion
    whitelist:
```

File: Resources/Prototypes/Entities/Objects/Devices/mousetrap.yml [hunk 5]
```diff
@@ -13,6 +13,9 @@
     - type: StepTrigger
       intersectRatio: 0.2
       requiredTriggeredSpeed: 0
+      triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Objects/Devices/mousetrap.yml:16]:
```suggestion
      triggerGroups:
```

File: Resources/Prototypes/Entities/Objects/Devices/pda.yml [hunk 6]
```diff
@@ -272,6 +272,9 @@
     paralyzeTime: 4
     launchForwardsMultiplier: 1.5
   - type: StepTrigger
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Objects/Devices/pda.yml:275]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Entities/Objects/Fun/dice.yml [hunk 7]
```diff
@@ -119,6 +119,9 @@
         mask:
         - ItemMask
   - type: StepTrigger
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Objects/Fun/dice.yml:122]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Entities/Objects/Materials/shards.yml [hunk 8]
```diff
@@ -64,6 +64,9 @@
       - !type:DoActsBehavior
         acts: [ "Destruction" ]
   - type: StepTrigger
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Objects/Materials/shards.yml:67]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Entities/Objects/Misc/land_mine.yml [hunk 9]
```diff
@@ -40,6 +40,9 @@
         params:
             maxDistance: 10
   - type: StepTrigger
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Objects/Misc/land_mine.yml:43]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Entities/Objects/Specific/Janitorial/soap.yml [hunk 10]
```diff
@@ -24,6 +24,9 @@
     paralyzeTime: 2
     launchForwardsMultiplier: 1.5
   - type: StepTrigger
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Objects/Specific/Janitorial/soap.yml:27]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Entities/Objects/Specific/Janitorial/soap.yml [hunk 11]
```diff
@@ -157,6 +160,9 @@
     paralyzeTime: 5
     launchForwardsMultiplier: 2.5
   - type: StepTrigger
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Objects/Specific/Janitorial/soap.yml:163]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Entities/Objects/Specific/Janitorial/soap.yml [hunk 12]
```diff
@@ -198,6 +204,9 @@
   - type: Slippery
     paralyzeTime: 2
   - type: StepTrigger
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Objects/Specific/Janitorial/soap.yml:207]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Entities/Tiles/bananium.yml [hunk 13]
```diff
@@ -47,6 +47,9 @@
     paralyzeTime: 2
     launchForwardsMultiplier: 1.5
   - type: StepTrigger
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Tiles/bananium.yml:50]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Entities/Tiles/chasm.yml [hunk 14]
```diff
@@ -14,6 +14,9 @@
     blacklist:
       tags:
       - Catwalk
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Tiles/chasm.yml:17]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Entities/Tiles/lava.yml [hunk 15]
```diff
@@ -13,6 +13,9 @@
     blacklist:
       tags:
         - Catwalk
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Tiles/lava.yml:16]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Entities/Tiles/liquid_plasma.yml [hunk 16]
```diff
@@ -13,6 +13,9 @@
     blacklist:
       tags:
       - Catwalk
+    triggerGroups: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Entities/Tiles/liquid_plasma.yml:16]:
```suggestion
    triggerGroups:
```

File: Resources/Prototypes/Nyanotrasen/Entities/Mobs/Species/felinid.yml [hunk 17]
```diff
@@ -80,6 +80,12 @@
     stripTimeReduction: 0
     stripTimeMultiplier: 0.667
   - type: StepTriggerImmune
+    whitelist: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Nyanotrasen/Entities/Mobs/Species/felinid.yml:83]:
```suggestion
    whitelist:
```

File: Resources/Prototypes/Traits/skills.yml [hunk 18]
```diff
@@ -270,6 +270,12 @@
   points: -3
   components:
     - type: StepTriggerImmune
+      whitelist: # WD EDIT
```
Comment (FoxxoTrystan) [inline Resources/Prototypes/Traits/skills.yml:273]:
```suggestion
      whitelist:
```

**Review by FoxxoTrystan — DISMISSED** [review]
diona need glass immunity (and slip but they already got (noslips).

infact, should noslip shoses just have also whitelists?

**Comment by VMSolidus** [comment]
> diona need glass immunity (and slip but they already got (noslips).
> 
> infact, should noslip shoses just have also whitelists?

Noslip is its own separate thing, and Diona have it. Also Diona have ShoesRequiredStepTriggerImmune so they don't care about glass.

**Comment by FoxxoTrystan** [comment]
Also that is some cursed CL

**Comment by VMSolidus** [comment]
> Also that is some cursed CL

Yep lol. But when porting from another codebase, we are supposed to preserve the original CL. I kinda like how they do it with bilingual changes. :)

**Comment by FoxxoTrystan** [comment]
its funny and  i understood it so, okay!
just dont forget diona

**Comment by DEATHB4DEFEAT** [comment]
Do the CL as is normally done

**Review by DEATHB4DEFEAT — APPROVED** [review]

**Review by FoxxoTrystan — CHANGES_REQUESTED** [review]
Not stale, please provide this to Dionas.
Before they had a special component or tag, witcth may got removed with the time, but again they are wood, sshould not be affected by glass.
```
- type: StepTriggerImmune
    whitelist:
      types:
      - Shard
```

**Comment by VMSolidus** [comment]
> Not stale, please provide this to Dionas. Before they had a special component or tag, witcth may got removed with the time, but again they are wood, sshould not be affected by glass.
> 
> ```
> - type: StepTriggerImmune
>     whitelist:
>       types:
>       - Shard
> ```

Done

**Review by FoxxoTrystan — APPROVED** [review]

**Review by OldDanceJacket — APPROVED** [review]

**Pull request merged**